File: 0874-walking-robot-simulation/0874-walking-robot-simulation.py
class Solution:
    def robotSim(self, commands: List[int], obstacles: List[List[int]]) -> int:
        direction = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        x, y, d = 0, 0, 0
        result = 0

        obs = set(map(tuple, obstacles))

        for command in commands:
            if command > 0:
                for i in range(command):
                    nx, ny = x + direction[d][0], y + direction[d][1]
                    if (nx, ny) in obs: break
                    x, y = nx, ny
                    result = max(result, x * x + y * y)
            elif command == -1:
                d = (d + 1) % 4
            else:
                d = (d - 1) % 4

        return result

File: 0874-walking-robot-simulation/0874-walking-robot-simulation.java
class Solution {
    public int robotSim(int[] commands, int[][] obstacles) {
        int[][] direction = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        int x = 0, y = 0, d = 0, result = 0;

        Set<String> obs = new HashSet<>();

        for (int[] obstacle: obstacles) {
            obs.add(obstacle[0] + "," + obstacle[1]);
        }

        for (int command: commands) {
            if (command > 0) {
                for (int i = 0; i < command; i++) {
                    int nx = x + direction[d][0], ny = y + direction[d][1];
                    if (obs.contains(nx + "," + ny)) {
                        break;
                    }
                    x = nx;
                    y = ny;
                    result = Math.max(result, x * x + y * y);
                }
            } else if (command == -1) {
                d = (d + 1) % 4;
            } else {
                d = (d + 3) % 4;
            }
        }

        return result;
    }
}